### PR TITLE
feat: bounded cache for informers

### PR DESCRIPTION
**Pull request by csviri**

### CODE REVIEW

**Review by metacosm — CHANGES_REQUESTED** [review]

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/cache/KubernetesResourceFetcher.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  @Override
+  public R fetchResource(String key) {
+    var resourceId = resourceIDFunction.apply(key);
+    return resourceId.getNamespace().map(ns -> client.resources(rClass).inNamespace(ns)
+        .withName(resourceId.getName()).get())
+        .orElse(client.resources(rClass).withName(resourceId.getName()).get());
+  }
+
+  /**
+   * This would not be needed
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/cache/KubernetesResourceFetcher.java:37]: this was a leftover, removed.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ResourceConfiguration.java [hunk 2]
```diff
@@ -122,4 +123,23 @@ default Set<String> getEffectiveNamespaces() {
     }
     return targetNamespaces;
   }
+
+  /**
+   * See <a href=
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ResourceConfiguration.java:128]: updated.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/cache/BoundedItemStore.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    return r -> Cache.namespaceKeyFunc(r.getMetadata().getNamespace(), r.getMetadata().getName());
+  }
+
+  protected R refreshMissingStateFromServer(String key) {
+    log.debug("Fetching resource from server for key: {}", key);
+    var newRes = resourceFetcher.fetchResource(key);
+    synchronized (this) {
+      log.debug("Fetched resource: {}", newRes);
+      if (newRes == null) {
+        existingMinimalResources.remove(key);
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/cache/BoundedItemStore.java:125]: will revisit this, came to my mind that there an case, where the current impl might be a problem of this method

**Review by metacosm — APPROVED** [review]

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/cache/BoundedItemStore.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/cache/BoundedItemStore.java]: made an update for this, to be absolutely safe

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/cache/BoundedItemStore.java [hunk 4]
```diff
@@ -121,14 +121,19 @@ protected R refreshMissingStateFromServer(String key) {
...
     synchronized (this) {
       log.debug("Fetched resource: {}", newRes);
+      var actual = cache.get(key);
       if (newRes == null) {
-        existingMinimalResources.remove(key);
+        // double-checking if actual, not received since.
+        // If received we just return. Since the resource from informer should be always leading,
+        // even if the fetched resource is null, this will be eventually received as an event.
+        if (actual == null) {
+          existingMinimalResources.remove(key);
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/cache/BoundedItemStore.java:130]: I still think that `actual` should be removed from `cache` as well.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ResourceConfiguration.java [hunk 5]
```diff
@@ -122,4 +123,23 @@ default Set<String> getEffectiveNamespaces() {
     }
     return targetNamespaces;
   }
+
+  /**
+   * Replaces the item store in informer. See underling <a href=
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ResourceConfiguration.java:128]:
```suggestion
   * Replaces the item store in informer. See underlying <a href=
```

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ResourceConfiguration.java [hunk 6]
```diff
@@ -122,4 +123,23 @@ default Set<String> getEffectiveNamespaces() {
...
+   * "https://github.com/fabric8io/kubernetes-client/blob/43b67939fde91046ab7fb0c362f500c2b46eb59e/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/informers/impl/DefaultSharedIndexInformer.java#L273">
+   * method</a> in fabric8 client informer implementation.
+   *
+   * The main goal, is to be able to use limited caches.
```
Comment (metacosm) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ResourceConfiguration.java:132]:
```suggestion
   * The main goal is to be able to use limited caches.
```

**Pull request merged**